File: client/src/modules/battle/logic/BattleMgr.js
var BattleMgr = cc.Layer.extend({
    ctor:function (
        randomSeed,
        listPlayerDeskCard,
        listPlayerDeskCardLevel,
        mapPlayer,
        listOpponentDeskCard,
        listOpponentDeskCardLevel,
        mapOpponent
    ) {
        this._super()

        this.randomCustom = new RandomCustom(randomSeed)

        // Add player game Mgr
        this.gamePlayerMgr = new GameMgr(listPlayerDeskCard, listPlayerDeskCardLevel, mapPlayer, true, new RandomCustom(randomSeed))
        this.gamePlayerMgr.initGamePlayerMgr()
        this.addChild(this.gamePlayerMgr, 0)

        // Add opponent game Mgr
        this.gameOpponentMgr = new GameMgr(listOpponentDeskCard, listOpponentDeskCardLevel, mapOpponent, false, new RandomCustom(randomSeed))
        this.gameOpponentMgr.initGameOpponentMgr()
        this.addChild(this.gameOpponentMgr, 0)

        // Queue action
        this.listActions = []

        // State
        this.randomCustom = new RandomCustom(randomSeed)
        this.listPreMonsterPlayer = []
        this.listPreMonsterOpponent = []
        this.ratioHPPlayer = 0
        this.ratioHPOpponent = 0
        this.countTurn = 0
        this.countLoop = 0
        this.timer = BATTLE.INIT_TIME_COUNT_DOWN

        this.delayFrameCreateMonster = 0
    },

    isTower: function (card) {
        return card.getType() == CARD.TOWER_TYPE
    },

    isMonster: function (card) {
        return card.getType() == CARD.MONSTER_TYPE
    },

    isPotion: function (card) {
        return card.getType() == CARD.POTION_TYPE
    },

    // Tao moi monster theo tung turn
    createTurnMonster: function () {
        this.delayFrameCreateMonster++
        if (this.delayFrameCreateMonster >= BATTLE.DELAY_FRAME_TO_CREATE_MONSTER ) {
            if (this.listPreMonsterPlayer.length > 0) {
                var monsterType = this.listPreMonsterPlayer.shift()
                this.gamePlayerMgr.createMonster(monsterType, 1, this.ratioHPPlayer, BATTLE.NORMAL_SHADOW, false)
            }

            if (this.listPreMonsterOpponent.length > 0) {
                var monsterType = this.listPreMonsterOpponent.shift()
                this.gameOpponentMgr.createMonster(monsterType, 1, this.ratioHPOpponent, BATTLE.NORMAL_SHADOW, false)
            }

            this.delayFrameCreateMonster = 0
        }
    },

    requestCreateMonsterImmediately: function () {
        if (this.gamePlayerMgr.listLogicMonster.length == 0
            && this.listPreMonsterPlayer.length == 0)
        {
            getBattleController().sendSummonSystemMonster();
        }
    },

    requestDropCard: function (coordinate, card) {
        if (this.isTower(card.info)) {
            return this.requestDropTowerPlayer(coordinate,card)
        }
        if (this.isMonster(card.info)) {
            return this.requestDropMonsterToOpponent(coordinate, card)
        }
        if (this.isPotion(card.info)) {
            return this.requestDropSpell(coordinate, card)
        }
    },

    // return: error code
    requestDropSpell: function (coordinate, card) {
        var statCard = CardStat.getInstance().getStatPotion(card.info.getIdInJsonFile())
        var mapTarget = statCard.map

        if (this.gamePlayerMgr.isInMap(coordinate)) {
            if (mapTarget == "enemy") {
                return gv.ERROR.SPELL_INVALID_POSITION_MAP
            } else {
                return this.gamePlayerMgr.requestDropSpell(this.gamePlayerMgr.currentEnergy, coordinate, card, true)
            }
        } else if (this.gameOpponentMgr.isInMap(coordinate)) {
            if (mapTarget == "player") {
                return gv.ERROR.SPELL_INVALID_POSITION_MAP
            } else {
                return this.gameOpponentMgr.requestDropSpell(this.gamePlayerMgr.currentEnergy, coordinate, card, false)
            }
        }
        return gv.ERROR.SPELL_INVALID_POSITION
    },

    // return: error code
    requestDropTowerPlayer: function (coordinate, card) {
        return this.gamePlayerMgr.requestDropTower(coordinate, card)
    },

    // return: error code
    requestDropMonsterToOpponent: function (coordinate, card) {
        return this.gameOpponentMgr.requestDropMonster(this.gamePlayerMgr.currentEnergy, coordinate, card)
    },

    createMonsterImmediately: function (){
        this.timer = BATTLE.INIT_TIME_COUNT_DOWN
        this.countTurn++
        if (this.countTurn <= BATTLE.MAX_TURN) {
            this.createListPreMonster()
            this.delayFrameCreateMonster = BATTLE.DELAY_FRAME_TO_CREATE_MONSTER
        }
    },

    handleDropMonster: function(action) {
        // Neu action do nguoi choi thuc hien
        if (action.userID == gv.user.id) {
            var card = this.gamePlayerMgr.deskCardMgr.listCard[action.indexCard]

            var sumLV = 0
            for (var i = 0; i < this.gamePlayerMgr.listLogicTower.length; i++) {
                sumLV += this.gamePlayerMgr.listLogicTower[i].evolution
            }
            var ratio;
            if (sumLV >= 30) {
                ratio = TURN_MONSTER.RATIO[30]
            } else {
                ratio = TURN_MONSTER.RATIO[sumLV]
            }

            this.gameOpponentMgr.dropMonster(card, ratio)
            this.gamePlayerMgr.updateEnergy(-card.getEnergy())

            gv.battleScene.updateDesk(action.indexCard)
        }
        // Neu action do doi thu thuc hien
        else {
            var card = this.gameOpponentMgr.deskCardMgr.listCard[action.indexCard]

            var sumLV = 0
            for (var i = 0; i < this.gameOpponentMgr.listLogicTower.length; i++) {
                sumLV += this.gameOpponentMgr.listLogicTower[i].evolution
            }
            var ratio;
            if (sumLV >= 30) {
                ratio = TURN_MONSTER.RATIO[30]
            } else {
                ratio = TURN_MONSTER.RATIO[sumLV]
            }

            this.gamePlayerMgr.dropMonster(card, ratio)
            this.gameOpponentMgr.updateEnergy(-card.getEnergy())

            // Chu y getNextIndex xong moi hideCard
            // -> Tranh truong hop index vong ve dau thi hien 1 card 2 lan
            this.gameOpponentMgr.deskCardMgr.getNextIndex()
            this.gameOpponentMgr.deskCardMgr.hideCard(action.indexCard)
        }
    },

    handleDropTower: function (action) {
        var gameMgr = action.userID == gv.user.id ? this.gamePlayerMgr : this.gameOpponentMgr
        var card = gameMgr.deskCardMgr.listCard[action.indexCard]

        gameMgr.dropTower(
            cc.p(action.coordinateX, action.coordinateY),
            card
        )

        // Neu action do nguoi choi thuc hien
        if (action.userID == gv.user.id) {
            gv.battleScene.updateDesk(action.indexCard)
        } else {
            // Chu y getNextIndex xong moi hideCard
            // -> Tranh truong hop index vong ve dau thi hien 1 card 2 lan
            this.gameOpponentMgr.deskCardMgr.getNextIndex()
            this.gameOpponentMgr.deskCardMgr.hideCard(action.indexCard)
        }
    },

    handleTargetObstacle: function (action) {
        let gameMgr = action.userID === gv.user.id ? this.gamePlayerMgr : this.gameOpponentMgr;
        if (action.errorCode === gv.ERROR.SUCCESS)
            gameMgr.forceTowerShootObstacle(gameMgr.getCell(action.coordinateX, action.coordinateY).obstacle);
    },

    handleCancelTower: function (action) {
        let gameMgr = action.userID === gv.user.id ? this.gamePlayerMgr : this.gameOpponentMgr;
        if (action.errorCode === gv.ERROR.SUCCESS)
            gameMgr.getCell(action.coordinateX, action.coordinateY).tower.destroy();
    },

    handleChangeTargetTower: function (action) {
        let gameMgr = action.userID === gv.user.id ? this.gamePlayerMgr : this.gameOpponentMgr;
        if (action.errorCode === gv.ERROR.SUCCESS)
            gameMgr.getCell(action.coordinateX, action.coordinateY).tower.setTargetMode(action.targetMode);
    },

    handleDropSpell: function (action) {
        var gameMgr = action.userID == gv.user.id ? this.gamePlayerMgr : this.gameOpponentMgr
        var card = gameMgr.deskCardMgr.listCard[action.indexCard]

        var isPlayerAction
        // Neu action do nguoi choi thuc hien
        if (action.userID == gv.user.id) {
            this.gamePlayerMgr.updateEnergy(-card.getEnergy())
            gv.battleScene.updateDesk(action.indexCard)
            isPlayerAction = true
        } else {
            this.gameOpponentMgr.updateEnergy(-card.getEnergy())

            // Chu y getNextIndex xong moi hideCard
            // -> Tranh truong hop index vong ve dau thi hien 1 card 2 lan
            this.gameOpponentMgr.deskCardMgr.getNextIndex()
            this.gameOpponentMgr.deskCardMgr.hideCard(action.indexCard)
            isPlayerAction = false
        }

        // Check xem drop spell tren map nao
        if (action.map == gv.user.id) {
            this.gamePlayerMgr.dropSpell(
                cc.p(action.coordinateX/1000, action.coordinateY/1000),
                card,
                isPlayerAction
            )
        } else {
            this.gameOpponentMgr.dropSpell(
                cc.p(action.coordinateX/1000, action.coordinateY/1000),
                card,
                isPlayerAction
            )
        }
    },

    handleCancelCard: function(action) {
        // Neu action do nguoi choi thuc hien
        if (action.userID == gv.user.id) {
            this.gamePlayerMgr.updateEnergy(-BATTLE.ENERGY_DESTROY_CARD)
            gv.battleScene.updateDesk(action.indexCard)
        }

        // Neu action do doi thu thuc hien
        else {
            this.gameOpponentMgr.updateEnergy(-BATTLE.ENERGY_DESTROY_CARD)

            // Chu y getNextIndex xong moi hideCard
            // -> Tranh truong hop index vong ve dau thi hien 1 card 2 lan
            this.gameOpponentMgr.deskCardMgr.getNextIndex()
            this.gameOpponentMgr.deskCardMgr.hideCard(action.indexCard)
        }
    },

    runListAction: function() {
        while (this.listActions.length > 0 && this.listActions[0].frameRunAction == this.countLoop){
            var action = this.listActions.shift()
            if (action.errorCode == gv.ERROR.SUCCESS) {
                switch (action.cmdID) {
                    case gv.CMD.SUMMON_SYSTEM_MONSTER:{
                        this.createMonsterImmediately();
                        break;
                    }
                    case gv.CMD.DROP_MONSTER:{
                        this.handleDropMonster(action);
                        break;
                    }
                    case gv.CMD.DROP_SPELL: {
                        this.handleDropSpell(action);
                        break;
                    }
                    case gv.CMD.DROP_TOWER:{
                        this.handleDropTower(action);
                        break;
                    }
                    case gv.CMD.TARGET_OBSTACLE: {
                        this.handleTargetObstacle(action);
                        break;
                    }
                    case gv.CMD.CANCEL_TOWER: {
                        this.handleCancelTower(action);
                        break;
                    }
                    case gv.CMD.CHANGE_TARGET_TOWER: {
                        this.handleChangeTargetTower(action);
                        break;
                    }
                    case gv.CMD.END_BATTLE:{
                        break;
                    }
                    case gv.CMD.CANCEL_CARD: {
                        this.handleCancelCard(action);
                        break;
                    }
                }
            } else {
                cc.log("Error battle: " + action.errorCode)
                if (action.userID == gv.user.id) {
                    gv.battleScene.showNotification(action.errorCode)
                }
            }
        }
    },

    updateCountLoop: function () {
        this.countLoop++
    },

    createListPreMonster: function () {
        // this.listPreMonsterPlayer = []
        // this.listPreMonsterOpponent = []

        var turn = this.countTurn

        // For player
        var sumLVOpponent = 0
        for (var i = 0; i < this.gameOpponentMgr.listLogicTower.length; i++) {
            sumLVOpponent += this.gameOpponentMgr.listLogicTower[i].evolution
        }
        var ratioPlayer;
        if (sumLVOpponent >= 30) {
            ratioPlayer = TURN_MONSTER.RATIO[30]
        } else {
            ratioPlayer = TURN_MONSTER.RATIO[sumLVOpponent]
        }
        this.ratioHPPlayer = ratioPlayer.hp

        // For opponent
        var sumLVPlayer = 0
        for (var i = 0; i < this.gamePlayerMgr.listLogicTower.length; i++) {
            sumLVPlayer += this.gamePlayerMgr.listLogicTower[i].evolution
        }
        var ratioOpponent;
        if (sumLVPlayer >= 30) {
            ratioOpponent = TURN_MONSTER.RATIO[30]
        } else {
            ratioOpponent = TURN_MONSTER.RATIO[sumLVPlayer]
        }
        this.ratioHPOpponent = ratioOpponent.hp

        for (var i = 0; i < TURN_MONSTER.TURN[turn].length; i++) {
            var monsterType = TURN_MONSTER.TURN[turn][i].type
            var indexInMonsterType = Math.floor(TURN_MONSTER.TYPE[monsterType].length * this.randomCustom.getRandom())
            var monster = TURN_MONSTER.TYPE[monsterType][indexInMonsterType]

            var amountPlayer, amountOpponent
            if (monsterType == TURN_MONSTER.BOSS_TYPE) {
                amountPlayer = 1
                amountOpponent = 1
            } else {
                amountPlayer = Math.ceil(TURN_MONSTER.TURN[turn][i].ratio * TURN_MONSTER.AMOUNT_OF_TYPE[monster] * ratioPlayer.amount)
                amountOpponent = Math.ceil(TURN_MONSTER.TURN[turn][i].ratio * TURN_MONSTER.AMOUNT_OF_TYPE[monster] * ratioOpponent.amount)
            }

            for (var j = 0; j < amountPlayer; j++) {
                this.listPreMonsterPlayer.push(monster)
            }

            for (var j = 0; j < amountOpponent; j++) {
                this.listPreMonsterOpponent.push(monster)
            }
        }
    },

    updateTimer: function (dt){
        this.timer -= dt
        if (this.timer < 0) {
            this.timer = BATTLE.INIT_TIME_COUNT_DOWN
            this.countTurn++
            if (this.countTurn <= BATTLE.MAX_TURN) {
                this.createListPreMonster()
                this.delayFrameCreateMonster = BATTLE.DELAY_FRAME_TO_CREATE_MONSTER
            }
        }
    },

    updateLogic: function (dt){
        this.updateTimer(dt)
        this.updateCountLoop()
        this.runListAction()
        this.createTurnMonster()
        this.gamePlayerMgr.updateLogic(dt)
        this.gameOpponentMgr.updateLogic(dt)

        // Ghi log ra file
        // LogState.getInstance().logToFile(this.gamePlayerMgr, this.countLoop)
    },

    updateUI: function (dt) {
        this.gamePlayerMgr.updateUI(dt)
        this.gameOpponentMgr.updateUI(dt)
    }
})

File: server/src/battle/BattleMgr.java
package battle;

import Constant.BattleError;
import battle.model.BattleAction;
import battle.model.Coordinate;
import battle.model.Obstacle;
import battle.model.RandomCustom;
import battle.model.shootableMonster.config.MonsterType;
import battle.model.shootableMonster.config.RatioMonsterTurn;
import battle.model.shootableMonster.config.TurnMonsterConfig;
import battle.model.tower.tower.Tower;
import battle.model.tower.towerAttack.TowerAttack;
import bitzero.server.entities.User;
import cmd.CmdDefine;
import cmd.request.battle.RequestBattleAction;
import config.battle.BattleConfig;
import config.battle.SynchronizeConfig;
import logState.LogState;
import model.battle.map.MapBattle;
import model.battle.map.Point;
import model.user.card.Card;
import util.battle.BattleUtils;

import java.util.*;

public class BattleMgr {
    public GameMgr gamePlayer1Mgr;
    public GameMgr gamePlayer2Mgr;
    public int countTurn;
    public int countLoop;
    public double timer;
    public int delayFrameCreateMonster;
    public int countReceiveAction;

    public ArrayDeque<BattleAction> queueAction1;
    public ArrayDeque<BattleAction> queueAction2;

    public ArrayDeque<Integer> listPreMonster1;
    public ArrayDeque<Integer> listPreMonster2;
    public int ratioHP1;
    public int ratioHP2;

    public List<BattleAction> listActionResponse;

    public User user1;
    public User user2;
    public RandomCustom randomCustom;

    public BattleMgr(int randomSeed, User user1, MapBattle mapBattle1, int[] listDeskCard1, User user2, MapBattle mapBattle2, int[] listDeskCard2) {
        this.user1 = user1;
        this.user2 = user2;

        randomCustom = new RandomCustom(randomSeed);

        gamePlayer1Mgr = new GameMgr(user1, mapBattle1, listDeskCard1, new RandomCustom(randomSeed));
        gamePlayer2Mgr = new GameMgr(user2, mapBattle2, listDeskCard2, new RandomCustom(randomSeed));

        countTurn = 0;
        countLoop = 0;
        timer = BattleConfig.INIT_TIME_COUNT_DOWN;

        delayFrameCreateMonster = 0;

        countReceiveAction= 0;

        queueAction1 = new ArrayDeque<>();
        queueAction2 = new ArrayDeque<>();

        listPreMonster1 = new ArrayDeque<>();
        listPreMonster2 = new ArrayDeque<>();
        ratioHP1 = 0;
        ratioHP2 = 0;

        listActionResponse = Collections.synchronizedList(new ArrayList<>());
    }

    public void addActionToQueue(ArrayDeque<BattleAction> queueAction, BattleAction battleAction) {
        if (queueAction.size() == SynchronizeConfig.MAX_USER_ACTION) {
            queueAction.removeFirst();
        }
        queueAction.addLast(battleAction);
    }

    public void addUserBattleAction(User user, int cmdID, RequestBattleAction actionArguments) {
        countReceiveAction++;
        BattleAction battleAction = new BattleAction(user, cmdID, countReceiveAction, actionArguments);
        if (this.user1.getId() == user.getId()) {
            addActionToQueue(queueAction1, battleAction);
        } else {
            addActionToQueue(queueAction2, battleAction);
        }
    }

    public void createListPreMonster() {
//        this.listPreMonster1.clear();
//        this.listPreMonster2.clear();

        int turn = this.countTurn;

        // For player 1
        int sumLV2 = 0;
        for (int i = 0; i < this.gamePlayer2Mgr.listTower.size(); i++) {
            sumLV2 += this.gamePlayer2Mgr.listTower.get(i).evolution;
        }
        RatioMonsterTurn ratio1;
        if (sumLV2 >= 30) {
            ratio1 = TurnMonsterConfig.RATIO[30];
        } else {
            ratio1 = TurnMonsterConfig.RATIO[sumLV2];
        }
        this.ratioHP1 = ratio1.hp;

        // For player 2
        int sumLV1 = 0;
        for (int i = 0; i < this.gamePlayer1Mgr.listTower.size(); i++) {
            sumLV1 += this.gamePlayer1Mgr.listTower.get(i).evolution;
        }
        RatioMonsterTurn ratio2;
        if (sumLV1 >= 30) {
            ratio2 = TurnMonsterConfig.RATIO[30];
        } else {
            ratio2 = TurnMonsterConfig.RATIO[sumLV1];
        }
        this.ratioHP2 = ratio2.hp;

        for (int i = 0; i < TurnMonsterConfig.TURN.get(turn).size(); i++) {
            int monsterType = TurnMonsterConfig.TURN.get(turn).get(i).type;
            int indexInMonsterType = (int)Math.floor(TurnMonsterConfig.TYPE[monsterType].length * this.randomCustom.getRandom());
            int monster = TurnMonsterConfig.TYPE[monsterType][indexInMonsterType];


            int amount1;
            int amount2;
            if (monsterType == TurnMonsterConfig.BOSS_TYPE) {
                amount1 = 1;
                amount2 = 1;
            } else {
                amount1 = (int) Math.ceil(
                        TurnMonsterConfig.TURN.get(turn).get(i).ratio
                                * TurnMonsterConfig.AMOUNT_OF_TYPE[monster]
                                * ratio1.amount
                );
                amount2 = (int) Math.ceil(
                        TurnMonsterConfig.TURN.get(turn).get(i).ratio
                                * TurnMonsterConfig.AMOUNT_OF_TYPE[monster]
                                * ratio2.amount
                );
            }

            for (int j = 0; j < amount1; j++) {
                this.listPreMonster1.addLast(monster);
            }

            for (int j = 0; j < amount2; j++) {
                this.listPreMonster2.addLast(monster);
            }
        }
    }

    // Tao moi monster theo tung turn
    public void createTurnMonster() {
        delayFrameCreateMonster++;
        if (delayFrameCreateMonster >= BattleConfig.DELAY_FRAME_TO_CREATE_MONSTER) {

            if (this.listPreMonster1.size() > 0) {
                int typeNumber = this.listPreMonster1.getFirst();
                this.listPreMonster1.removeFirst();

                MonsterType monsterType = TurnMonsterConfig.CAST_TO_MONSTER_TYPE[typeNumber];

                gamePlayer1Mgr.createMonster(monsterType, 1, this.ratioHP1, false);
            }

            if (this.listPreMonster2.size() > 0) {
                int typeNumber = this.listPreMonster2.getFirst();
                this.listPreMonster2.removeFirst();

                MonsterType monsterType = TurnMonsterConfig.CAST_TO_MONSTER_TYPE[typeNumber];

                gamePlayer2Mgr.createMonster(monsterType, 1, this.ratioHP2, false);
            }

            delayFrameCreateMonster = 0;
        }
    }

    public void createMonsterImmediately(){
        timer = BattleConfig.INIT_TIME_COUNT_DOWN;
        countTurn++;
        if (countTurn <= BattleConfig.MAX_TURN) {
            createListPreMonster();
            delayFrameCreateMonster = BattleConfig.DELAY_FRAME_TO_CREATE_MONSTER;
        }
    }

    public void updateTimer(double dt){
        timer -= dt;
        if (timer < 0) {
            timer = BattleConfig.INIT_TIME_COUNT_DOWN;
            countTurn++;
            if (countTurn <= BattleConfig.MAX_TURN) {
                createListPreMonster();
                delayFrameCreateMonster = BattleConfig.DELAY_FRAME_TO_CREATE_MONSTER;
            }
        }
    }

    public void updateCountLoop() {
        countLoop++;
    }

    public void handleSummonSystemMonster(BattleAction battleAction) {
        GameMgr gameMgr = battleAction.user.getId() == user1.getId() ? gamePlayer1Mgr : gamePlayer2Mgr;
        ArrayDeque<Integer> listPreMonster = battleAction.user.getId() == user1.getId() ? listPreMonster1 : listPreMonster2;

        if (gameMgr.listMonster.size() == 0 && listPreMonster.size() == 0) {
            createMonsterImmediately();
        } else {
            battleAction.errorCode = BattleError.SUMMON_EXIST_MONSTER_ON_MAP;
        }
        listActionResponse.add(battleAction);
    }

    public void handleCancelCard(BattleAction battleAction) {
        GameMgr gameMgr = battleAction.user.getId() == user1.getId() ? gamePlayer1Mgr : gamePlayer2Mgr;

        // Loi index card
        if (!gameMgr.deskCardMgr.isShowed[battleAction.actionArguments.indexCard]){
            battleAction.errorCode = BattleError.CANCEL_CARD_INVALID_INDEX_CARD;
        }
        // Loi khong du energy
        else if (gameMgr.currentEnergy < BattleConfig.ENERGY_DESTROY_CARD) {
            battleAction.errorCode = BattleError.CANCEL_CARD_NOT_ENOUGH_ENERGY;
        }
        // Cancel card
        else {
            // Update energy
            gameMgr.updateEnergy(-BattleConfig.ENERGY_DESTROY_CARD);

            // Update desk card (cancel card)
            // Note: getNextIndex xong moi hideCard (Error: index vong ve dau thi hien 1 card 2 lan)
            gameMgr.deskCardMgr.getNextIndex();
            gameMgr.deskCardMgr.hideCard(battleAction.actionArguments.indexCard);
        }
        listActionResponse.add(battleAction);
    }

    public void handleDropMonster(BattleAction battleAction) {
        GameMgr gameMgr = battleAction.user.getId() == user1.getId() ? gamePlayer1Mgr : gamePlayer2Mgr;
        GameMgr gameMgrOpponent = battleAction.user.getId() == user1.getId() ? gamePlayer2Mgr : gamePlayer1Mgr;
        Card card = gameMgr.deskCardMgr.listCard[battleAction.actionArguments.indexCard];

        // Check index card
        if (!gameMgr.deskCardMgr.isShowed[battleAction.actionArguments.indexCard]){
            battleAction.errorCode = BattleError.MONSTER_INVALID_CARD_USING;
        }
        // Check energy
        else if (gameMgr.currentEnergy < card.getEnergy()) {
            battleAction.errorCode = BattleError.MONSTER_NOT_ENOUGH_ENERGY;
        }
        else {
            // Get ratio
            int sumLV = 0;
            for (int i = 0; i < gameMgr.listTower.size(); i++) {
                sumLV += gameMgr.listTower.get(i).evolution;
            }
            RatioMonsterTurn ratio;
            if (sumLV >= 30) {
                ratio = TurnMonsterConfig.RATIO[30];
            } else {
                ratio = TurnMonsterConfig.RATIO[sumLV];
            }

            // Drop monster on opponent map
            gameMgrOpponent.dropMonster(card, ratio);

            // Update energy
            gameMgr.updateEnergy(-card.getEnergy());

            // Update desk card
            // Note: getNextIndex xong moi hideCard (Error: index vong ve dau thi hien 1 card 2 lan)
            gameMgr.deskCardMgr.getNextIndex();
            gameMgr.deskCardMgr.hideCard(battleAction.actionArguments.indexCard);
        }
        listActionResponse.add(battleAction);
    }

    public void handleDropSpell(BattleAction battleAction) {
        User remainingUser = battleAction.user.getId() == user1.getId() ? user2 : user1;
        GameMgr gameMgr = battleAction.user.getId() == user1.getId() ? gamePlayer1Mgr : gamePlayer2Mgr;
        GameMgr gameMgrOpponent = battleAction.user.getId() == user1.getId() ? gamePlayer2Mgr : gamePlayer1Mgr;
        Card card = gameMgr.deskCardMgr.listCard[battleAction.actionArguments.indexCard];

        // Check index card
        if (!gameMgr.deskCardMgr.isShowed[battleAction.actionArguments.indexCard]){
            battleAction.errorCode = BattleError.SPELL_INVALID_CARD_USING;
        }
        // Check energy
        else if (gameMgr.currentEnergy < card.getEnergy()) {
            battleAction.errorCode = BattleError.SPELL_NOT_ENOUGH_ENERGY;
        } else {
            // Update energy
            gameMgr.updateEnergy(-card.getEnergy());

            // Neu spell duoc drop tren map user goi
            if (battleAction.actionArguments.map == 1) {
                battleAction.errorCode = gameMgr.dropSpell(
                        new Coordinate((double)battleAction.actionArguments.coordinateX/1000,
                                (double)battleAction.actionArguments.coordinateY/1000),
                        card,
                        battleAction.user.getId()
                );
                battleAction.actionArguments.map = battleAction.user.getId();
            }
            // Neu spell duoc drop tren map opponent cua client goi
            else {
                battleAction.errorCode = gameMgrOpponent.dropSpell(
                        new Coordinate((double)battleAction.actionArguments.coordinateX/1000,
                                (double)battleAction.actionArguments.coordinateY/1000),
                        card,
                        battleAction.user.getId()
                );
                battleAction.actionArguments.map = remainingUser.getId();
            }

            if (battleAction.errorCode == BattleError.SUCCESS) {
                // Update desk card
                // Note: getNextIndex xong moi hideCard (Error: index vong ve dau thi hien 1 card 2 lan)
                gameMgr.deskCardMgr.getNextIndex();
                gameMgr.deskCardMgr.hideCard(battleAction.actionArguments.indexCard);
            }
        }

        // Add action to list response
        listActionResponse.add(battleAction);
    }

    public void handleDropTower(BattleAction battleAction) {
        GameMgr gameMgr = battleAction.user.getId() == user1.getId() ? gamePlayer1Mgr : gamePlayer2Mgr;
        Card card = gameMgr.deskCardMgr.listCard[battleAction.actionArguments.indexCard];

        // Drop tower
        battleAction.errorCode = gameMgr.dropTower(
            new Point(battleAction.actionArguments.coordinateX, battleAction.actionArguments.coordinateY),
            card,
            battleAction.actionArguments.indexCard
        );

        if (battleAction.errorCode == BattleError.SUCCESS) {
            // Update desk card
            // Note: getNextIndex xong moi hideCard (Error: index vong ve dau thi hien 1 card 2 lan)
            gameMgr.deskCardMgr.getNextIndex();
            gameMgr.deskCardMgr.hideCard(battleAction.actionArguments.indexCard);
        }

        // Add action to list response
        listActionResponse.add(battleAction);
    }

    public void handleTargetObstacle(BattleAction battleAction) {
        try {
            GameMgr gameMgr = battleAction.user.getId() == user1.getId() ? gamePlayer1Mgr : gamePlayer2Mgr;
            Obstacle obstacle = gameMgr.getCell(battleAction.actionArguments.coordinateX, battleAction.actionArguments.coordinateY).obstacle;
            if (obstacle == null || obstacle.isDied)
                throw new NullPointerException();
            gameMgr.forceTowerShootObstacle(obstacle);
        } catch (NullPointerException e) {
            battleAction.errorCode = BattleError.TARGET_OBSTACLE_INVALID_POSITION;
        } finally {
            // Add action to list response
            this.listActionResponse.add(battleAction);
        }
    }

    public void handleCancelTower(BattleAction battleAction) {
        try {
            GameMgr gameMgr = battleAction.user.getId() == user1.getId() ? gamePlayer1Mgr : gamePlayer2Mgr;
            Tower tower = gameMgr.getCell(battleAction.actionArguments.coordinateX, battleAction.actionArguments.coordinateY).tower;
            if (tower.disable)
                throw new NullPointerException();
            tower.destroy();
        } catch (NullPointerException e) {
            battleAction.errorCode = BattleError.CANCEL_TOWER_INVALID_POSITION;
        } finally {
            // Add action to list response
            this.listActionResponse.add(battleAction);
        }
    }

    public void handleChangeTargetTower(BattleAction battleAction) {
        try {
            GameMgr gameMgr = battleAction.user.getId() == user1.getId() ? gamePlayer1Mgr : gamePlayer2Mgr;
            Tower tower = gameMgr.getCell(battleAction.actionArguments.coordinateX, battleAction.actionArguments.coordinateY).tower;
            if (tower.disable)
                throw new NullPointerException();
            ((TowerAttack) tower).setTargetMode(battleAction.actionArguments.targetMode);
        } catch (NullPointerException e) {
            battleAction.errorCode = BattleError.CHANGE_TARGET_TOWER_INVALID_POSITION;
        } catch (IllegalStateException e) {
            battleAction.errorCode = BattleError.CHANGE_TARGET_TOWER_INVALID_TARGET_MODE;
        } catch (ClassCastException e) {
            battleAction.errorCode = BattleError.CHANGE_TARGET_TOWER_INVALID_TOWER_TYPE;
        } finally {
            // Add action to list response
            this.listActionResponse.add(battleAction);
        }
    }

    public void runAction(BattleAction battleAction) {
        try {
            switch (battleAction.cmdID) {
                case CmdDefine.SUMMON_SYSTEM_MONSTER:{
                    handleSummonSystemMonster(battleAction);
                    break;
                }
                case CmdDefine.DROP_MONSTER:{
                    handleDropMonster(battleAction);
                    break;
                }
                case CmdDefine.DROP_SPELL: {
                    handleDropSpell(battleAction);
                    break;
                }
                case CmdDefine.DROP_TOWER: {
                    handleDropTower(battleAction);
                    break;
                }
                case CmdDefine.TARGET_OBSTACLE: {
                    handleTargetObstacle(battleAction);
                    break;
                }
                case CmdDefine.END_BATTLE: {
                    break;
                }
                case CmdDefine.CANCEL_TOWER: {
                    handleCancelTower(battleAction);
                    break;
                }
                case CmdDefine.CHANGE_TARGET_TOWER: {
                    handleChangeTargetTower(battleAction);
                    break;
                }
                case CmdDefine.CANCEL_CARD:{
                    handleCancelCard(battleAction);
                    break;
                }
            }
        } catch (Exception e) {
            System.out.println(e);
        }
    }

    public void handleAction() {
        while (queueAction1.size() + queueAction2.size() > 0){
            int countReceiveAction1 = SynchronizeConfig.MAX_COUNT_RECEIVE_ACTION;
            int countReceiveAction2 = SynchronizeConfig.MAX_COUNT_RECEIVE_ACTION;

            if (!queueAction1.isEmpty()) {
                countReceiveAction1 = queueAction1.getFirst().countReceiveAction;
            }
            if (!queueAction2.isEmpty()) {
                countReceiveAction2 = queueAction2.getFirst().countReceiveAction;
            }
            if (countReceiveAction1 < countReceiveAction2) {
                runAction(queueAction1.getFirst());
                queueAction1.removeFirst();
            } else {
                runAction(queueAction2.getFirst());
                queueAction2.removeFirst();
            }
        }
    }

    public boolean checkEndBattle() {
        return gamePlayer1Mgr.currentHP <= 0 || gamePlayer2Mgr.currentHP <= 0 || this.countTurn > BattleConfig.MAX_TURN;
    }

    private int getResult() {
        if (gamePlayer1Mgr.currentHP > 0 && gamePlayer2Mgr.currentHP > 0 && this.countTurn <= BattleConfig.MAX_TURN)
            return BattleConfig.RESULT_UNKNOW;
        if (gamePlayer1Mgr.currentHP == gamePlayer2Mgr.currentHP)
            return BattleConfig.RESULT_DRAW;
        return gamePlayer1Mgr.currentHP > gamePlayer2Mgr.currentHP ? BattleConfig.RESULT_WINNER_PLAYER1 : BattleConfig.RESULT_WINNER_PLAYER2;
    }

    public int getResultPlayer(int player) {
        int result = getResult();
        if (result == BattleConfig.RESULT_UNKNOW || result == BattleConfig.RESULT_DRAW)
            return BattleConfig.RESULT_DRAW;
        return result == player ? BattleConfig.RESULT_WIN : BattleConfig.RESULT_LOSE;
    }

    public void update(double dt, boolean doHandleAction){
        updateTimer(dt);

        if (checkEndBattle())
            return;

        updateCountLoop();
        if (doHandleAction) {
            handleAction();
        }
        createTurnMonster();
        gamePlayer1Mgr.update(dt);
        gamePlayer2Mgr.update(dt);

        // Ghi log ra file
//        LogState.getInstance().logToFile(this.gamePlayer1Mgr, this.countLoop);
//        LogState.getInstance().logToFile(this.gamePlayer2Mgr, this.countLoop);
    }
}